### PR TITLE
Upgrade arkworks to 0.4

### LINKED ISSUE #1289: Update arkworks to 0.4
There have been plenty of changes in arkworks and we need to update to the recently released version.

### PR DESCRIPTION
**Pull request by paberr**
## What's in this pull request?
- Upgrade arkworks to 0.4
- Generalize some of the circuits
- Remove `Vec<bool>` in favour of actual types and `Vec<u8>`
- Prepare pedersen hash to be replaced by arkworks equivalent
- Fix optimization level for `zkp-circuits`

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

**Comment by hrxi** [comment]
This implicitly upgrades all dependencies, including e.g. libmdbx to 0.1.12, which we apparently intentionally didn't do: https://github.com/nimiq/core-rs-albatross/pull/1238#pullrequestreview-1246000563.

File: four_validators.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[[validator]]
```
Comment (jsdanielh) [inline four_validators.toml]: Was this file included by mistake? If we want to include this test scenario we could add it into `.github/devnet_topologies/`

File: rpc-interface/Cargo.toml [hunk 2]
```diff
@@ -17,8 +17,8 @@ is-it-maintained-open-issues = { repository = "nimiq/core-rs" }
 maintenance = { status = "experimental" }
 
 [dependencies]
-ark-groth16 = { version = "0.3", default-features = false }
-ark-mnt6-753 = "0.3"
+ark-groth16 = "0.4"
```
Comment (jsdanielh) [inline rpc-interface/Cargo.toml]: Do we actually need `parallel` feature (that would be the default feature we would be pulling now)?

File: zkp-circuits/Cargo.toml [hunk 3]
```diff
@@ -19,22 +19,28 @@ log = { package = "tracing", version = "0.1", features = ["log"] }
...
-ark-mnt4-753 = "0.3"
-ark-mnt6-753 = "0.3"
-ark-relations = "0.3"
-ark-r1cs-std = "0.3.1"
-ark-serialize = "0.3"
-ark-std = "0.3"
+ark-crypto-primitives = { version = "0.4", features = ["crh", "prf", "r1cs"] }
+ark-ec = "0.4"
+ark-ff = "0.4"
+ark-groth16 = "0.4"
```
Comment (jsdanielh) [inline zkp-circuits/Cargo.toml]: Do we actually need `parallel` feature (that would be the default feature we would be pulling now)?

File: rpc-interface/Cargo.toml (hunk 2 above)
Comment (paberr) [reply rpc-interface/Cargo.toml]: It is definitely extremely useful for proving. We might not want to enable it for the web client, though.

File: four_validators.toml (hunk 1 above)
Comment (ii-cruz) [reply four_validators.toml]: it was a mistake, we have removed it

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**